### ISSUE TITLE
v3: helm template does not render NOTES.txt

### ISSUE THREAD

**Issue by vainikkaj** [helm/helm#6901]
Helm version:
```
$ h3 version
version.BuildInfo{Version:"v3.0.0-rc.3", GitCommit:"2ed206799b451830c68bff30af2a52879b8b937a", GitTreeState:"clean", GoVersion:"go1.13.4"}
```

Steps to recreate:
```
$ h3 create foo
Creating foo
```

```
$ h3 template foo -s templates/NOTES.txt --notes
Error: unknown flag: --notes
```

```
$ tree foo
foo
├── charts
├── Chart.yaml
├── templates
│   ├── deployment.yaml
│   ├── _helpers.tpl
│   ├── ingress.yaml
│   ├── NOTES.txt
│   ├── serviceaccount.yaml
│   ├── service.yaml
│   └── tests
│       └── test-connection.yaml
└── values.yaml
```

How feature works in Helm 2.x 
 ```
$ helm template bar -x templates/NOTES.txt --notes
---
# Source: bar/templates/NOTES.txt
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=bar,app.kubernetes.io/instance=release-name" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:80
```

**Comment by hickeyma** [comment]
@vainikkaj Good catch. This looks like it might have been a regression. It seems to have been removed in #5365. @bacongobbler Was this by design or unintentionally?

**Comment by hickeyma** [comment]
@bacongobbler If unintentional, I'll push a PR for it.

**Comment by bacongobbler** [comment]
this might've been an unintentional regression.

**Comment by hickeyma** [comment]
ok, will push a pr later or tomorrow morning

**Comment by bacongobbler** [comment]
Hey @hickeyma, do you think you'll have time to look into this before we cut 3.0? If not we should probably move this over to 3.1.

**Comment by vainikkaj** [comment]
I'm keeping my fingers crossed for 3.0 fix :smiley:

**Comment by hickeyma** [comment]
@bacongobbler Sorry but due to  some unexpected events, I have unable to get push it. After talking on slack, I will leave it to you as you said you might get it in.

**Comment by mattfarina** [comment]
I tried to fix this but found a nuance we need to talk about first. This is why I'm moving it to 3.1.0.

In Helm v2 when running `helm template` with the `--notes` flag it generates the notes for that chart and all sub-charts. Each is separated like the YAML parts are and the individual source is noted.

In Helm v2 when running `helm install` with the `--render-subchart-notes` all of the notes are combined into one output. For example, if you installed `stable/wordpress` using that flag the mariadb and wordpress notes would be back to back with no separator. Personally, I find this to be a confusing UX. I only noticed it now because I don't tend to use the `--render-subchart-notes` flag.

In Helm v3 the rendering process was based around the install process rather than a separate flow like Helm v2. In a deeply nested part of Helm the notes are generated and separated from the other manifests. If the sub-chart notes are rendered they are all combined the way helm v2 install presents them.

This means we have lost the ability to render each of the notes output as a separate file for `helm template`. The feature will not be exactly the same for v3.

I have some questions I would like feedback on...

* What, if any, are the impacts of rendering one notes output instead of each of the subcharts separately?
* Should there be headings between each of the charts/sub-charts when rendering notes? This would affect both `helm template` and `helm install` (and other commands)
* On `helm template` should the sub-chart notes be hidden behind an addition flag like they are on `helm install`?

Feedback on this is appreciated so that we can make sure the experience provided by showing notes works moving forward.

**Comment by technosophos** [comment]
Given this, I think @mattfarina is right in delaying this to 3.1. Trying to rush something out the door without thinking through the UX would lock us into an anit-pattern from now until Helm 4, since fixing it would require a breaking change.

**Comment by bacongobbler** [comment]
We haven't heard of any activity/feedback from the OP on this ticket, so I'm going to remove this from the 3.1.0 milestone until further feedback to the concerns raised from the other maintainers have been addressed. Thanks!

**Comment by ttauveron** [comment]
As a temporary workaround, you can run the following command :

``` shell
helm install chart --dry-run  --generate-name
```
This will render the NOTES.txt file.

**Comment by brendarearden** [comment]
https://github.com/rancher/rancher/issues/26999 we attempted to use the work around provided by @ttauveron but since --dry-run doesn't support CRDs we are still having difficulty displaying notes for all templates - would like to see `--notes` support added back in

**Comment by AceHack** [comment]
Yes would also like to see this, do we know when it can be fixed for template?

**Comment by AceHack** [comment]
Bump

**Comment by jglick** [comment]
> ```shell
> helm install chart --dry-run  --generate-name
> ```

Unfortunately this requires an actual cluster, unlike `helm template`.

**Comment by jglick** [comment]
Can we punt on subcharts and even the `--notes` option and just make

```bash
helm template $chart -s templates/NOTES.txt
```

render that file, so you can test or iteratively develop complex substitutions in it without needing a live cluster?

**Comment by lhankinsgdl** [comment]
I'll note that 

> helm install chart --dry-run  --generate-name

...seems to also require that the chart not be installed.
(I got `Error: rendered manifests contain a resource that already exists.`)

**Comment by eshepelyuk** [comment]
Hi any news on this feature ?

**Comment by abdennour** [comment]
actually we want to DRY: readme of a chart should refer to NOTES.txt

**Comment by cmoulliard** [comment]
The trick proposed to render the `NOTES.txt` is not working using helm 3.5.4 with a local helm charts
```
Current folder ~/myproject/helm/rbac

ls -la templates   
total 56
drwxr-xr-x  9 cmoullia  staff   288 Apr 30 11:41 .
drwxr-xr-x  9 cmoullia  staff   288 Apr 30 11:42 ..
-rw-r--r--  1 cmoullia  staff   196 Apr 30 11:35 NOTES.txt
-rw-r--r--  1 cmoullia  staff  1752 Apr 30 10:45 _helpers.tpl
-rw-r--r--  1 cmoullia  staff   859 Apr 30 11:17 cluster-role-binding.yaml
...

helm template rbac ./ -s templates/NOTES.txt
Error: could not find template templates/NOTES.txt in chart

```

**Comment by jglick** [comment]
@cmoulliard if you meant https://github.com/helm/helm/issues/6901#issuecomment-770070832, that was just a proposal—indeed it does not work today.

**Comment by stephenpope** [comment]
Did fixing this feature get abandoned ?

**Comment by jinleileiking** [comment]
Any solution?

**Comment by Dentrax** [comment]
Bump

**Comment by kennedy** [comment]
Bump

**Comment by consideRatio** [comment]
It is a relevant feature, but I think there is a need to help out drive development about it.

## Current behavior of `--render-subchart-notes`

Its relevant to overview the current behavior if we consider changing it. I conclude that `--render-subchart-notes` is silently ignored when used in the `helm template`, but for `helm install` it is functioning correctly and as documented.

The documentation says `if set, render subchart notes along with the parent`, so practically, the documentation is also currently only valid for the `helm install` situation as the `along with the parent` part isn't true anyhow for when used with `helm template`.

#### Steps to create a dummy chart for debugging

```shell
# create helm chart "a" with nested charts
# a has subchart b and c
# b has subchart d
cd /tmp
helm create a && echo "a FINDME" > a/templates/NOTES.txt
cd a/charts
helm create b && echo "b FINDME" > b/templates/NOTES.txt
helm create c && echo "c FINDME" > c/templates/NOTES.txt
cd b/charts
helm create d && echo "d FINDME" > d/templates/NOTES.txt
cd ../../..
```

#### Reproduction of broken `helm template --render-subchart-notes .`

Nothing is rendered when `--render-subchart-notes` is used with `helm template`.

```
# check for any NOTES.txt being rendered
helm template --render-subchart-notes . | grep FINDME
```

#### Output from `helm install --dry-run --render-subchart-notes test .`

Note that the ordering of these, the ordering seems random.

```
<CHART A TEMPLATES>
NOTES:
<NOTES FROM CHART B>
<NOTES FROM CHART A>
<NOTES FROM CHART D>
<NOTES FROM CHART C>
```

## What UX makes sense?

In https://github.com/helm/helm/issues/6901#issuecomment-553010477 @bacongobbler asked for input on what UX to go for.

#### My short-term UX suggestion

- Add `--render-notes` flag for `helm template` and `helm install` alongside `--render-subchart-notes` to toggle if the root chart's notes should be rendered. Let `--render-notes` be false by default for `helm template`, but true by default for `helm install`, to match current behavior.
- Update the help text for `--render-subchart-notes` note about its mention of `along with the parent`, something that should be controlled by `render-notes` now.
- For now, stick with current behavior from `helm install` with how notes are to be rendered for `helm template`. The current behavior is to render notes after all other templates has been rendered, with the separator line `NOTES:`. The notes are rendered one after the other in a random order without additional new-line spacing or titles.

#### My long-term UX suggestion

I think we should go for the short-term suggestion above at first and not go straight to this, as this involves breaking changes etc.

When `helm template` is used, we can see the following separators between templates.

- I think we should treat NOTES.txt as any other rendered template with separators that include a mention of where they come from.
  ```
  ---
  # Source: a/templates/tests/test-connection.yaml
  ```
- I think NOTES.txt templates should be rendered last.
- I think NOTES.txt from the main chart and its subcharts should be ordered deterministically when rendered, so that if we have a structure like a,b,c,d above the ordering would be a>b>d>c where d comes before c because it is a dependency of b which evaluates in full before c.

## Now what?

- Everyone: please provide feedback to the short-term UX proposal above.
- Maintainers: please help with yes/no decisions - is there a proposal available that you encourage work towards at this point in time?
- Maintainers: please help with who-does-what decisions - do you wish for someone in the community to work towards this, or do you wish to do it yourself?

**Comment by jglick** [comment]
> we should treat NOTES.txt as any other rendered template

If you mean to do so by default, this could break tooling which expects the output of `helm template` to be (multidocument) YAML. Would be safer to render it as comments (prepending `# ` to every line):

```yaml
---
# Source: xxx/templates/service-account.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: xxx
  namespace: xxx
---
# Source: xxx/templates/NOTES.txt
# Here is how you use this chart:
# Run: la la la
```

which seems to be tolerated

```console
$ helm template xxx | yq e -ojson | jq -s
[
  {
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": {
      "name": "xxx",
      "namespace": "xxx"
    }
  },
  null
]
```

**Comment by PowerSurge1** [comment]
Whatever we do here, let's consider https://github.com/argoproj/argo-cd/issues/4733 as a use case. It looks like ArgoCD uses 'helm template' and then applies the result, rather than executing helm directly. Looks like both sides want the NOTES.TXT available.

**Comment by Dentrax** [comment]
Any roadmap planned here?

**Comment by sandipchitale** [comment]
I also wish there was a way to **generate** some manifests for any Kubernetes Resources or CRDs during regular install/upgrade/ rollback, but **not actually deploy** them. Maybe templates with special extension like .generate.yml. These will capture the exact set of .Values and .Capabilities and .Release objects. These could then be referenced in Notes.txt file. The output could go to some output directories that capture the release name and number  as seen in ```helm history```  and install/upgrade/rollback flavor so that user knows which install/upgrade/rollback generated them. Such generated manifests could be invoked by the user later to run one-off jobs etc later to make some changes in deployed release but use consistently generated manifests.

**Comment by sandipchitale** [comment]
Actually, I just realized that can declare a manifest template in NOTES.txt and it will not be deployed but will be rendered:

```
NOTES.txt

Other notes stuff

---
# Release Name:      {{ .Release.Name }}
# Release Namespace: {{ .Release.Namespace }}
# Release Revision:  {{ .Release.Revision }}
# Release Install:   {{ .Release.IsInstall }}
# Release Upgrade:   {{ .Release.IsUpgrade }}
---
# templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: {{ include "gen.fullname" . }}
  labels:
    {{- include "gen.labels" . | nindent 4 }}
spec:
  type: {{ .Values.service.type }}
  ports:
    - port: {{ .Values.service.port }}
      targetPort: http
      protocol: TCP
      name: http
  selector:
    {{- include "gen.selectorLabels" . | nindent 4 }}
```

**Comment by consideRatio** [comment]
https://github.com/helm/helm/issues/6901#issuecomment-553010477 requested input on what UX to go for: this is a _non-breaking_ and proposal that addresses this issue and fits well with future breaking changes.

## Proposal: add `--render-notes` flag to `helm [template|install]`
  - `--render-notes` determines if top level chart NOTES.txt is rendered.
  - `--render-subchart-notes` determines if subchart NOTES.txt are rendered.
  - For backward compatibility, `--render-notes` is by default made false for `helm template` and true for `helm install`.

**Comment by eshepelyuk** [comment]
@consideRatio what about `helm upgrade` ?

**Comment by alfsch** [comment]
For me this issue is still open

**Comment by jglick** [comment]
Yes it was closed simply due to lack of activity.

It seems https://github.com/helm/helm/blob/21e694e0f9575700f16f6c87002c6ccdc434edc4/.github/workflows/stale-issue-bot.yaml#L9 has not been updated to pick up https://github.com/actions/stale/issues/789 which would at least let it show up in grey rather than purple. Maybe #11498 would fix that.

**Comment by davinkevin** [comment]
+1 on this one. We have no way to test the code generated by `NOTES.txt` without a cluster connected (because `install|upgrade` requires a cluster, even if `--dry-run` and `--debug` are set)

**Comment by mohittalele** [comment]
any update on this issue ?

**Comment by mnp** [comment]
One more complexity here - running `helm test` always prints the NOTES. Is that intentional?  We view it as clutter and should be optional.  Maybe that's another +1 for adding some `--render-notes=true|false` UX?

**Comment by abdennour** [comment]
after 5 years. any update ?

**Comment by abdennour** [comment]
adding api-version explicitly on `helm template`, then manage it in your helm chart templates... it will work 

.i.e.
```sh
helm template ....  --api-versions "apps.openshift.io/v1"
```

Credits to Purestorage / Portworx team!

**Comment by aabouzaid** [comment]
> adding api-version explicitly on `helm template`, then manage it in your helm chart templates... it will work
> 
> .i.e.
> 
> ```shell
> helm template ....  --api-versions "apps.openshift.io/v1"
> ```
> 
> Credits to Purestorage / Portworx team!

Which Helm CLI version works with that? I've tried it and it doesn't work with `v3.16.3`.

**Comment by ivankatliarchuk** [comment]
Any updates? We can open a pull request?

**Comment by daviddob** [comment]
Also looking for this same functionality - being able to render notes would be extremely helpful for gitops workflows like argocd + flux where we are templating and then applying resultant manifests to the clusters

**Comment by Bhargav-InfraCloud** [comment]
Hi team,
Based on the comments on this issue, I've created the following PRs. PTAL. Thanks!
1. https://github.com/helm/helm/pull/31096
~2. https://github.com/helm/helm/pull/31097~ No new features to v3.

PS: The commit size has become too big due to tests and related files. The actual changes aren't that big.

**Comment by TerryHowe** [comment]
This won't make v3, so I changed the label to v4

**Comment by scottrigby** [comment]
`--render-subcharts-notes` does not belong on `helm template`. It didn’t make sense in Helm 3, it doesn’t make sense in Helm 4, and keeping it around only preserves confusion, not compatibility.

`helm template` has a very clear contract: render Kubernetes manifests. Its output is expected to be machine-consumable YAML, often piped directly into tools like kubectl or kustomize.

`NOTES.txt` is explicitly outside that contract. It exists to print human-readable guidance after `helm install` or `helm upgrade`. While it supports templating, it is not intended to produce Kubernetes resources.

These two concerns don’t overlap.

If the goal is to preview NOTES output, Helm already supports that via `helm install` / `helm upgrade --dry-run` (client or server), including subchart notes. For existing releases, `helm get notes` has long covered this use case.

The Rancher workaround mentioned earlier is obsolete and no longer reflects how Helm is used today.

The ArgoCD use case mentioned earlier is not applicable to `helm template`. As Argo maintainer said in https://github.com/argoproj/argo-cd/issues/4733#issuecomment-721389036:
> How does `helm template` produce the output of notes since it's only supposed to produce apply-able YAML?

So the issue here isn’t that Helm lacks a way to preview NOTES output, it's that a flag with no meaningful behavior still shows up in `helm template` help, implying a use case that doesn’t actually exist. As such, this should be closed as not a bug / wontfix, with a follow-up to deprecate and hide the flag from `helm template`.

**Comment by ErikEngerd** [comment]
As an alternative I am now using [gotpl](https://github.com/belitre/gotpl) which provides direct go templating. The code of that github project is very simple, if you would need more helm functions it would not be difficult to extend. 

Generating NOTES.txt can be done using a values.yaml or on the CLI. For instance: 

```
gotpl mychart-dir/templates/NOTES.txt  -f mychart-dir/values.yaml  \
  --set Values.hostAliases[0].ip=1.2.3.4 \
  --set Values.hostAliases[0].hostnames[0]=example.com
```
This would effectively use values.yaml with overrides

```
hostAliases:
  - ip: 1.2.3.4
    hostnames:
       - example.com 
```
